### ISSUE TITLE
No BBU Information from hpssacli

### ISSUE THREAD

**Issue by jkcgn** [glensc/nagios-plugin-check_raid#132]
Hi Glen, 

i know the bbu-monitoring option is still experimental, but if i use the hpssacli plugin i should at least get an "OK", because the command "/usr/sbin/hpssacli controller all show status" already contains this information:

root@jk:/usr/lib/nagios/plugins/raidnew# ./check_raid -p hpssacli --bbu-monitoring
OK: hpssacli:[Smart Array P840: Array A(OK)[LUN1:OK]]

root@jk:/usr/lib/nagios/plugins/raidnew# ./check_raid -d -p hpssacli --bbu-monitoring
DEBUG EXEC: /usr/sbin/hpssacli controller all show status at ./check_raid line 474.
DEBUG EXEC: /usr/sbin/hpssacli controller slot=3 logicaldrive all show at ./check_raid line 474.
OK: hpssacli:[Smart Array P840: Array A(OK)[LUN1:OK]]

root@jk:/usr/lib/nagios/plugins/raidnew# /usr/sbin/hpssacli controller all show status
Smart Array P840 in Slot 3
   Controller Status: OK
   Cache Status: OK
   Battery/Capacitor Status: OK

root@jk:/usr/lib/nagios/plugins/raidnew# /usr/sbin/hpssacli controller slot=3 logicaldrive all show
Smart Array P840 in Slot 3
array A
logicaldrive 1 (7.3 TB, RAID 6, OK)

Any Ideas?

thx,
Jan

**Comment by glensc** [comment]
there's just no handling of the BBU information: [Plugins/hpacucli.pm](https://github.com/glensc/nagios-plugin-check_raid/blob/4.0.0/lib/App/Monitoring/Plugin/CheckRaid/Plugins/hpacucli.pm)

you can start with by providing commands and command outputs, so i can add them to repository and somebody can work on that.

other than that, `cciss` plugin which [is recommended](https://github.com/glensc/nagios-plugin-check_raid/issues/114#issuecomment-138866801) for such hardware, does not also have bbu support?

**Comment by jkcgn** [comment]
Yes, i know that cciss it recommended and meanwhile cciss_vol_status also supports the hpsa driver and the old ccis block driver, so would be perfect for my hardware. 

But cciss can´t retreive any BBU Information. Thats the reason i would like to use hpssacli instead of cciss.

I also read about the kernel lock problem with hpacucli. But AFAIK hpacucli doesnt support 
kernel > 2.6 (without a uname26 wrapper script) and it seems that HP moved to hpssacli and the locking problem is gone meanwhile.

I could provide some more bash commands and output if needed, but i think the relevant command and command output to get the BBU status from hpssacli, was the one i already wrote:

`root@jk:/usr/lib/nagios/plugins/raidnew# /usr/sbin/hpssacli controller all show status
Smart Array P840 in Slot 3
   Controller Status: OK
   Cache Status: OK
   Battery/Capacitor Status: OK
`

Maybe for a first solution, it would be enough to grep the string after "Battery/Capacitor Status" and add this to the output of check_raid. And if it is not "OK", the nagios status should be "Warning".

If you need a more detailed command to grep the battery status i can provide it, but unfortunately i cant write in perl.

**Comment by jkcgn** [comment]
Ok, forget what i wrote about hpssacli and the resolved kernel locking yesterday! Today i had one, because of monitoring with hpssacli -  I'll use cciss_vol_status now.
And maybe the BBU Information with cciss_vol_status is missing, because the controller uses CacheVault and BBU, even if this should be supported in fresh versions.

**Comment by glensc** [comment]
to be able to work on this, you need to provide test commands and data:

https://github.com/glensc/nagios-plugin-check_raid/issues/132#issuecomment-185333931

**Comment by nmacgreg** [comment]
Thanks, Glen, great code, we've been using this for years!

So, the magic command that spits out battery status is: 
```
[root@milton ~]# hpssacli controller all show status

Smart Array P440ar in Slot 0 (Embedded)
   Controller Status: OK
   Cache Status: OK
   Battery/Capacitor Status: OK


```
(Yes, that ends with two blank lines!)
Yes, this is the command that's already being run!
Yes, it's also reporting the status of the cache-RAM, and hmmm, I don't see that reported "OK" in the output, either (I've used megacli before, I know it does this by default, without --bbu-monitoring.)

What for following da rulez, DEBUG, first with : 
```
[root@milton ~]# /usr/lib64/nagios/plugins/check_raid -d -p hpssacli --bbu-monitoring
check_raid 4.0.6
Visit <https://github.com/glensc/nagios-plugin-check_raid#reporting-bugs> how to report bugs

DEBUG EXEC: /sbin/hpssacli controller all show status at /usr/lib64/nagios/plugins/check_raid line 482.
DEBUG EXEC: /sbin/hpssacli controller slot=0 logicaldrive all show at /usr/lib64/nagios/plugins/check_raid line 482.
OK: hpssacli:[Smart Array P440ar[OK]: Array A(OK)[LUN1:OK], Array B(OK)[LUN2:OK]]
```
... and without --bbu-monitoring:
```
root@milton ~]# /usr/lib64/nagios/plugins/check_raid -d -p hpssacli
check_raid 4.0.6
Visit <https://github.com/glensc/nagios-plugin-check_raid#reporting-bugs> how to report bugs

DEBUG EXEC: /sbin/hpssacli controller all show status at /usr/lib64/nagios/plugins/check_raid line 482.
DEBUG EXEC: /sbin/hpssacli controller slot=0 logicaldrive all show at /usr/lib64/nagios/plugins/check_raid line 482.
OK: hpssacli:[Smart Array P440ar[OK]: Array A(OK)[LUN1:OK], Array B(OK)[LUN2:OK]]
```
Let me know if I can help further / test. 
Context: CentOS7 on DL560 Gen9; no cciss driver in the kernel :(, otherwise I would use the recommended cciss_vol_status.   And, hpssacli-2.40-13.0.x86_64, fresh from the HP repo.
